### PR TITLE
[WIP] Ambient on-demand XDS for DNS

### PR DESCRIPTION
**Pull request by nmittler**
Modifies the ambient index to support lookup of both services and workloads by hostname.

### CODE REVIEW

**Comment by nmittler** [comment]
Will add tests after https://github.com/istio/istio/pull/45797 lands.

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go [hunk 1]
```diff
@@ -159,18 +216,20 @@ func (a *AmbientIndex) updateWaypoint(sa model.WaypointScope, addr *workloadapi.
 	return updates
 }
 
-// All return all known workloads. Result is un-ordered
+// All return all known workloads and services. Result is un-ordered
 func (a *AmbientIndex) All() []*model.AddressInfo {
 	a.mu.RLock()
 	defer a.mu.RUnlock()
 	res := make([]*model.AddressInfo, 0, len(a.byPod)+len(a.serviceByAddr)+len(a.byWorkloadEntry))
+	for _, services := range a.serviceByHostname {
```
Comment (nmittler) [inline pilot/pkg/serviceregistry/kube/controller/ambientindex.go:224]: NOTE: this was needed to support headless services, which don't have an address.

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go [hunk 2]
```diff
@@ -86,6 +87,48 @@ func (c *Controller) generatePodUID(p *v1.Pod) string {
...
 }
 
+// servicesByNamespacedHostname returns the service with the given namespace and hostname. If
+// the provided namespace is a wildcard ("*"), all services with the hostname are returned.
+func (a *AmbientIndex) servicesByNamespacedHostname(namespace, hostname string) []*model.ServiceInfo {
+	out := make([]*model.ServiceInfo, 0)
+	if services := a.serviceByHostname[hostname]; services != nil {
+		if namespace == "*" {
+			// The query was for all namespaces. Return all services with the hostname.
+			out = append(out, services...)
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/ambientindex.go:97]: nit: return return this directly?

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:97]: I suppose that makes it possible to be modified (probably accidentally) later though. so LGTM

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go [hunk 3]
```diff
@@ -86,6 +87,48 @@ func (c *Controller) generatePodUID(p *v1.Pod) string {
...
+				}
+			}
+		}
+	}
+	return out
+}
+
+// workloadsByNamespacedHostname returns the workload with the given namespace and hostname. If
+// the provided namespace is a wildcard ("*"), all workloads with the hostname are returned.
+func (a *AmbientIndex) workloadsByNamespacedHostname(namespace, hostname string) []*model.WorkloadInfo {
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/ambientindex.go:113]: Hmm... I think workload hostname is globally unique and maybe not need the namespace part. Maybe not though. Is it only headless services?

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go (hunk 3 above)
Comment (nmittler) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:113]: Yeah I wasn't sure if there could be cases where workload hostname could be scoped to a namespace, same as services. I'm happy to remove the namespace component for now to keep it simple.

**Pull request closed without merging**